gossip/tx_scrambler.go

```diff
@@ -0,0 +1,119 @@
+package gossip
+
+import (
+	"bytes"
+	"github.com/ethereum/go-ethereum/common"
+	"slices"
+)
+
+type scramblerEntry struct {
+	hash   common.Hash
+	sender common.Address
+	nonce  uint64
+}
+
+// getExecutionOrder first removes any entries with duplicate hashes, then sorts the list by XORed hashes.
+// Furthermore, if there are entries with same sender, these entries are sorted by their nonce (lower comes first).
+func getExecutionOrder(entries []*scramblerEntry) []*scramblerEntry {
+	uniqueList, salt, hasDuplicateAddresses := analyseEntryList(entries)
+	scrambleTransactions(uniqueList, salt)
+
+	// no need to sort more
+	if !hasDuplicateAddresses {
+		return uniqueList
+	}
+
+	return sortTransactionsByNonce(uniqueList)
+}
+
+// sortTransactionsByNonce finds any duplicate senders and sorts their transactions by nonce ascending.
+func sortTransactionsByNonce(entries []*scramblerEntry) []*scramblerEntry {
```

Comment:
This operation is not deterministic. See for example this test case:
```

func TestTxScrambler_SortTransactionsByNonce_IsDeterministic(t *testing.T) {
	entries := []*scramblerEntry{}
	for i := range 10 {
		entries = append(entries, &scramblerEntry{
			hash: common.Hash{byte(i)},
		})
	}

	printOrder := func(entries []*scramblerEntry) string {
		builder := strings.Builder{}
		for _, entry := range entries {
			builder.WriteString(fmt.Sprintf("%d ", entry.hash[0]))
		}
		return builder.String()
	}

	want := sortTransactionsByNonce(entries)
	for range 10 {
		input := slices.Clone(want)
		shuffleEntries(input)
		got := sortTransactionsByNonce(input)
		if !reflect.DeepEqual(want, got) {
			t.Errorf("sort is not deterministic, want %v, got %v",
				printOrder(want),
				printOrder(got),
			)
		}
	}
}
```

Reply:
Actually, forget that. I was writing this up too fast, I have to rethink this test case.